alibi_detect/od/pytorch/pca.py

```diff
@@ -0,0 +1,237 @@
+
+        Returns
+        -------
+        The principle components of the reference data.
+        """
+        x -= self.x_ref_mean
+        cov_mat = (x.t() @ x)/(len(x)-1)
+        _, V = torch.linalg.eigh(cov_mat)
+        return V[:, :-self.n_components]
+
+    def compute_score(self, x: torch.Tensor) -> torch.Tensor:
+        """Compute the outlier score.
+
+        Centers the data and projects it onto the principle components. The score is then the sum of the
+        squared projections.
+
+        Parameters
+        ----------
+        x
+            The test data.
+
+        Returns
+        -------
+        The outlier score.
+        """
+        x_cen = x - self.x_ref_mean
+        x_pcs = x_cen @ self.pcs
+        return (x_pcs**2).sum(1)
+
+
+class KernelPCATorch(PCATorch):
+    def __init__(
+            self,
+            n_components: int,
+            kernel: Optional[Callable],
+            device: Optional[Union[str, torch.device]] = None
+            ):
+        """Kernel variant of the PyTorch backend for PCA detector.
+
+        Parameters
+        ----------
+        n_components:
+            The number of dimensions in the principle subspace.
+        kernel
+            Kernel function to use for outlier detection.
+        device
+            Device type used. The default tries to use the GPU and falls back on CPU if needed. Can be specified by
+            passing either ``'cuda'``, ``'gpu'``, ``'cpu'`` or an instance of ``torch.device``.
+        """
+        super().__init__(device=device, n_components=n_components)
+        self.kernel = kernel
+
+    def compute_pcs(self, x: torch.Tensor) -> torch.Tensor:
+        """Compute the principle components of the reference data.
+
+        We compute the principle components of the reference data using the kernel matrix and then
+        return the last `n_components` of the eigenvectors. These correspond to the invariant dimensions
+        of the data. Changes in these dimensions are used to compute the outlier score.
+
+        Parameters
+        ----------
+        x
+            The reference data.
+
+        Returns
+        -------
+        The principle components of the reference data.
+        """
+        K = self.compute_kernel_mat(x)
+        D, V = torch.linalg.eigh(K)
+        pcs = V / torch.sqrt(D)[None, :]
+        return pcs[:, -self.n_components:]
+
+    def compute_score(self, x: torch.Tensor) -> torch.Tensor:
+        """Compute the outlier score.
+
+        Centers the data and projects it onto the principle components. The score is then the sum of the
+        squared projections.
+
+        Parameters
+        ----------
+        x
+            The test data.
+
+        Returns
+        -------
+        The outlier score.
+        """
+        k_xr = self.kernel(x, self.x_ref)
+        k_xr_row_sums = k_xr.sum(1)
+        _, n = k_xr.shape
+        k_xr_cen = k_xr - self.k_col_sums[None, :]/n - k_xr_row_sums[:, None]/n + self.k_sum/(n**2)
```

Comment:
This should be

m, n = k_xr.shape
k_xr_cen = k_xr - self.k_col_sums[None, :]/m - k_xr_row_sums[:, None]/n + self.k_sum/(m*n)

My bad - I see this was a typo in my original demo

Reply:
Ah, good point. Although I think we've got `m` and `n` the wrong way around above.

Reply:
Yes, good spot!